engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java

```diff
@@ -93,40 +72,42 @@ public Either<Failure, DecisionEvaluationResult> evaluateDecision(
     }
 
     final var decisionId = decisionIdOrFailure.get();
-    // todo(#8571): avoid parsing drg every time
-    final var decisionOrFailure = findDecisionById(decisionId);
+    final var decisionOrFailure = decisionBehavior.findDecisionById(decisionId);
+    final Either<Failure, ParsedDecisionRequirementsGraph> drgOrFailure =
+        decisionOrFailure.flatMap(decision -> decisionBehavior.findAndParseDrgByDecision(decision));
+    if (drgOrFailure.isLeft()) {
+      // any failures above have the same error type and the correct scope
+      final Failure formattedFailure =
+          decisionBehavior.formatDecisionLookupFailure(drgOrFailure.getLeft(), decisionId);
```

Comment:
💭 Instead of making this behavior responsible for the formatting of the `DecisionBehavior` Failure message we could consider to just do this formatting in the methods of there behavior directly. We could do that by adding a `mapLeft` to the finding and parsing of the decision/drg

```java
  public Either<Failure, PersistedDecision> findDecisionById(final String decisionId) {
    return Either.ofOptional(
            decisionState.findLatestDecisionById(BufferUtil.wrapString(decisionId)))
        .orElse(new Failure("no decision found for id '%s'".formatted(decisionId)))
        .mapLeft(failure -> formatDecisionLookupFailure(failure, decisionId));
  }

  public Either<Failure, ParsedDecisionRequirementsGraph> findAndParseDrgByDecision(
      final PersistedDecision persistedDecision) {
    return findDrgByDecision(persistedDecision)
        .flatMap(drg -> parseDrg(drg.getResource()))
        .mapLeft(
            failure ->
                formatDecisionLookupFailure(
                    failure, BufferUtil.bufferAsString(persistedDecision.getDecisionId())));
  }
```

Alternatively, I suppose we could also get rid of the `formatDecisionLookupFailure` method and just put it in the `BpmnDecisionBehavior`. Let's see if we will need it for standalone evaluation. Maybe we want a different failure message here? Probably not, so we'd just be delaying finding a clean solution 😄 

Just some random thoughts, I can definitely live with the way it is currently.

Reply:
That's a nice idea. I'll do it. :+1: 

Working with the `Either` took some getting used to, especially the part that a `Left` just passes through a `flatMap` and a `Right` through a `mapLeft`.

That influenced how I worked with the failures. I didn't think of re-using the `formatDecisionLookupFailure` in each method because I was obsessed with avoiding duplication. :smile: